### PR TITLE
#6 :sparkles: backup migrated files

### PR DESCRIPTION
**Pull request by d-ryan-ashcraft**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -179,6 +179,28 @@ The configurations file name to look for in the classpath (all matches will be u
 
 Default: `migrations.json`
 
+### backupOriginalMigratedFiles
+Controls if backups of original files prior to migration are saved.
+
+Default: `true`
+
+### backupCustomLocation
+Customizes the location where backup of original files prior to migration is performed.  By default, the system temp
```
Comment (d-ryan-ashcraft) [inline README.md]: A (Administrative): backup*s* ... *are stored*

File: README.md [hunk 2]
```diff
@@ -179,6 +179,28 @@ The configurations file name to look for in the classpath (all matches will be u
...
+greater control.
+
+Default: `system temp directory`
+
+### numberOfBackupsToKeep
+The number of backups for original files prior to migration to keep.  Once this number is hit, the backup files will 
+rotate, with the oldest being purged.  
+
+It is worth noting that this controls number of old files kept, not the total number files kept.  For instance, if set 
+to 5, you will have 5 *old* backup files as well as one active backup.  This is based on the file rotation library works.
```
Comment (d-ryan-ashcraft) [inline README.md]: A: *how*

File: baton-maven-plugin/src/test/resources/specifications/backup-original-files-before-migration.feature [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    Given a file to migrate "<fileName>"
+    When the migration backup is executed <numberOfExecutions> and <backupLimit>
+    Then one copy of the file can be found in the project's target location
+    And <backupLimit> plus one copies of the file can be found in the project's backup location
+
+    Examples:
+      | fileName  | numberOfExecutions | backupLimit |
+      | fileE.txt | 2                  | 1           |
+      | fileF.txt | 3                  | 2           |
+      | fileF.txt | 10                 | 9           |
```
Comment (d-ryan-ashcraft) [inline baton-maven-plugin/src/test/resources/specifications/backup-original-files-before-migration.feature]: S (Substantive): add some examples that show many more executes that the backup limit still result in just the backup limit + 1

File: baton-maven-plugin/src/main/java/org/technologybrewery/baton/AbstractMigration.java [hunk 4]
```diff
@@ -86,6 +112,67 @@ public MigrationSummary execute(FileSet[] fileSets) {
...
+                String persistentBackupPathBase;
+                if (StringUtils.isNotBlank(backupCustomLocation)) {
+                    persistentBackupPathBase = getBasedir().getAbsolutePath() + "/" + backupCustomLocation;
+                } else {
+                    persistentBackupPathBase = FileUtils.getTempDirectoryPath();
+                }
+
+                String originalPathAsString = Paths.get(originalFile.toURI()).toString();
+                String baseDirectoryFilePath = Paths.get(getBasedir().toURI()).toString();
+                String localPath = originalPathAsString.replace(baseDirectoryFilePath, "");
```
Comment (epwilkins) [inline baton-maven-plugin/src/main/java/org/technologybrewery/baton/AbstractMigration.java:127]: A: The Paths API has a `relativize`/`resolve` method that may be more readable.  May also be less readable given we have a mix of `File` and `Path` here.

Comment (d-ryan-ashcraft) [reply baton-maven-plugin/src/main/java/org/technologybrewery/baton/AbstractMigration.java:127]: Yeah - my first pass used it, but ended up being less clear and required more processing to remove the "../../.." content once the value returned.

**Review by epwilkins — APPROVED** [review]

**Pull request merged**